onnx/shape_inference.py

```diff
@@ -131,4 +131,23 @@ def infer_node_outputs(
     return {key: onnx.TypeProto.FromString(out) for key, out in outputs.items()}
 
 
+def infer_function_output_types(
```

Comment:
The name is longer, but I prefer that since it is more precise. `infer_function_type`, in my mind, sounds different. To me it suggests inferring a type-signature for the function. The key difference is that we are doing the inference for a specific calling-context (as opposed to inferring a type that is valid for all calling-contexts). Hard to convey the distinction, but the current name is the best I can come up with.